File: day23/crab_cups.py
#!/usr/bin/env python3
from sys import argv

class LinkedList:
    def __init__(self, val):
        self.val = val
        self.next_node = None

def create_circle_list(cups):
    d = {}
    a = [int(c) for c in cups]
    for i in a:
        d[i] = LinkedList(i)

    for i in range(len(a)):
        d[a[i]].next_node = d[a[(i+1) % len(a)]]

    return d[a[0]]        

def do_move(clist):
    # bypass the next 3 nodes
#    print('bypassing')
    front = clist.next_node
    back = front.next_node.next_node
    clist.next_node = back.next_node

    # which vals were just picked?
#    print('picked?')
    picked = set([front.val, front.next_node.val, front.next_node.next_node.val])
#    print(f'{picked=}')

    # find the destination node num
#    print('find destination num')
    dest = clist.val - 1
    if dest == 0:
        dest = 9
    while dest in picked:
        dest = dest - 1
        if dest == 0:
            dest = 9
#    print(f'{dest=}')

    # find the destination node
#    print('find destination node')
    p = clist.next_node
    while p.val != dest:
        p = p.next_node

    # insert the picked nodes at p
#    print('insert')
    back.next_node = p.next_node
    p.next_node = front

def compute_score(clist):
    s = ''

    # find 1
    while clist.val != 1:
        clist = clist.next_node

    # find all the non-1s
    while True:
        clist = clist.next_node
        if clist.val == 1:
            break
        else:
            s += str(clist.val)
    return s

clist = create_circle_list(argv[1])

for move in range(100):
    do_move(clist)
    clist = clist.next_node
#    print('cur =', clist.val)
    print(f'{move=}  ', end='')
    tmp = clist
    for _ in range(len(argv[1])):
        print(f'{tmp.val} ', end='')
        tmp = tmp.next_node
    print()

# compute score
print('Part 1:', compute_score(clist))

File: day23/crab_cups.cpp
#include <stdlib.h>
#include <string.h>
#include <iostream>
#include "LinkedList.h"

using namespace std;

void usage() {
    cout << "crab_cubs N MOVES [12] INPUT" << endl;
    exit(0);
}

LinkedList *create_circle_list(const unsigned long long n, const char *init) {
    LinkedList *cups = new LinkedList[n];
    const size_t initlen = strlen(init);
    
    // initialize values
    for (unsigned long long i = 0; i < n; i++)
        cups[i].set_val(i+1);

    // link all but the last digit in init
    for (size_t i = 0; i < initlen - 1; i++)
        cups[init[i]- '0' - 1].set_next(cups + init[i+1] - '0' - 1);

    // link from initlen to n-1
    for (size_t i = initlen; i < n; i++)
        cups[i].set_next(cups + i + 1);
                                   
    // now complete the circle
    if (n == strlen(init))
        cups[init[n-1] - '0' - 1].set_next(cups + init[0] - '0' - 1);
    else {
        cups[init[strlen(init)-1] - '0' - 1].set_next(cups + initlen);
        cups[n-1].set_next(cups + init[0] - '0' - 1);
    }

    return cups;
}

void do_move(LinkedList *cups, LinkedList *cup, const unsigned long long N) {
    // bypass the next 3 nodes
    LinkedList *front = cup->next();
    LinkedList *back = front->next()->next();
    cup->set_next(back->next());

    // which values were just picked?
    unsigned long long picked[3] = {front->val(), front->next()->val(), back->val()};

    // find the destination value
    unsigned long long dest = cup->val() - 1;
    if (dest == 0)
        dest = N;

    while (1) {
        bool found = false;
        for (int i = 0; i < 3; i++) {
            if (picked[i] == dest) {
                dest--;
                if (dest == 0)
                    dest = N;
                found = true;
                break;
            }
        }
        if (!found)
            break;
    }

    // find the destination node
    LinkedList *p = cups + dest - 1;

    // insert the picked notes at p
    back->set_next(p->next());
    p->set_next(front);
}

void part1(LinkedList *cups) {
    LinkedList *p = cups->next();
    int score = 0;
    for (int i = 0; i < 8; i++) {
        score *= 10;
        score += p->val();
        p = p->next();
    }

    cout << "Part 1: " << score << endl;
}

void part2(LinkedList *cups) {
    cout << "Part 2: " << cups->next()->val() * cups->next()->next()->val() << endl;
}

int main(int argc, char *argv[]) {
    if (argc != 5) {
        cout << "argc == " << argc << endl;
        usage();
    }
    
    const unsigned long long N = atoi(argv[1]);
    const int MOVES = atoi(argv[2]);
    const int part = atoi(argv[3]);
    const char *init = argv[4];

    LinkedList *cups = create_circle_list(N, init);
    LinkedList *cup = cups + init[0] - '0' - 1;
    for (int i = 0; i < MOVES; i++) {
        do_move(cups, cup, N);
        cup = cup->next();
    }

    if (part == 1)
        part1(cups);
    else if (part == 2)
        part2(cups);
}
